### DISCUSSION TITLE
Przesyłanie json'a i pliku w jednym requeście, a Django REST Framework

### DISCUSSION THREAD

**Discussion by itchystone** [emkarcinos/WMIAdventure#428] [Codding]
Już od dłuższego czasu myślę, jak dołożyć obrazek/ikonę karty do requesta z utworzeniem nowej karty. Przesyłanie obrazu w jsonie nie jest takie proste i chyba będę musiał wysyłać *form data* na frontendzie, a na backendzie będę musiał dopisać `Parser`, żeby to ogarnąć. Opisane jest to tutaj na stacku [link](https://stackoverflow.com/questions/20473572/django-rest-framework-file-upload/50514022#50514022)

**Comment by itchystone** [comment]
Można by też rozbić ten jeden request na dwa:
- pierwszy z danymi karty
- drugi z dosłanym obrazem

https://www.trell.se/blog/file-uploads-json-apis-django-rest-framework/

Nawet preferowałbym tą opcję

**Reply by emkarcinos** [reply]
Ten post na stacku jest dosyć stary i z tego co widzę to w najnowszym DRF to jest jakoś już rozwiązane [tutaj piszą że działa z marszu](https://stackoverflow.com/questions/20303252/django-rest-framework-imagefield).

Jeżeli jednak okażę się, że coś jest z tym nie tak, to zrobiłbym jednak ten parser raz i mielibyśmy z głowy przesyłanie plików do końca projektu. To będzie kluczowe, bo w trybie Adventure będzie dużo plików (zdjęcia typu tła, postacie, itp będą praktycznie cały czas wysyłane do klienta)

Potestuję sam czy to w ogóle działa i napiszę tutaj co mi wychodzi i co działa

**Reply by itchystone** [reply]
Próbowałem jeszcze, ale nie działa. Na [dokumentacji drf](https://www.django-rest-framework.org/api-guide/fields/#file-upload-fields) jest napisane, że jsonowy parser nie obsługuje uploadowania plików:

![image](https://user-images.githubusercontent.com/43492764/128177069-b2372bc3-fa3a-4a5d-9342-5c8ca846fc67.png)

**Reply by emkarcinos** [reply]
Przetestowałem to i mi wszystko działa.
Zrobiłem PUT do karty która była już w bazie i dodałem jej ten obraz. Uploaduje się poprawnie i można go wyświetlić normalnie.
W POST nie da się stworzyć całego JSONa z plikiem, więc można robić po prostu POST nowej karty i potem od razu robić PUT wysyłając ten plik.

**Reply by emkarcinos** [reply]
Zrobiłem nawet requesta do serwera testowego i wszystko śmiga:
![image](https://user-images.githubusercontent.com/11946540/128258446-05bbf253-1472-4a64-8f9d-05e697bd1fef.png)

![image](https://user-images.githubusercontent.com/11946540/128258309-6f811889-ea8d-4402-a741-373ee365767e.png)

**Reply by emkarcinos** [reply]
Posiedzę jednak trochę nad tym taskiem żeby zrobić jakieś generowane nazwy plików, bo takie randomowe nazwy będą kiepsko wyglądały

**Reply by itchystone** [reply]
Okej, to ja zrobie PUT
